examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -41,101 +50,346 @@ def freq_mask(freqs, fmin, fmax):
+       
         self.estimator = estimator
         self.n_components = n_components
-
+        self.rank = rank
+        self.sampling_freq=sampling_freq      
+        self.cov_method_params = cov_method_params
+        
+    
     def fit(self, inst):
         """Fit"""
-        if self.picks == None:
-            self.picks_ = mne.pick_types(inst.info, meg=True, eeg=False, ref_meg=False)
-        else:
-            self.picks_ = self.picks
-        if isinstance(inst, mne.io.base.BaseRaw):
+        
+        
+        if isinstance(inst, BaseRaw):
+            if self.picks_ is None:
+                raise ValueError('picks should be provided')
+            self.max_components=len(self.picks_)
             inst_signal = inst.copy()
             inst_signal.filter(picks=self.picks_, **self.filt_params_signal)
-
+            self.Xs=inst_signal
+            #noise
             inst_noise = inst.copy()
             inst_noise.filter(picks=self.picks_, **self.filt_params_noise)
-            if self.subtract_signal:
-                inst_noise._data[self.picks_] -= inst_signal._data[self.picks_]
-            cov_signal = mne.compute_raw_covariance(
-                inst_signal, picks=self.picks_, method=self.estimator)
-            cov_noise = mne.compute_raw_covariance(inst_noise, picks=self.picks_,
-                method=self.estimator)
+            # subtract signal:
+            inst_noise._data[self.picks_] -= inst_signal._data[self.picks_]
+           
+            cov_signal = compute_raw_covariance(
+                inst_signal, picks=self.picks_, method=self.estimator, rank=self.rank)
+            cov_noise = compute_raw_covariance(inst_noise, picks=self.picks_,
+                method=self.estimator, rank=self.rank)
             del inst_noise
             del inst_signal
         else:
-            raise NotImplementedError()
+            if isinstance(inst, BaseEpochs)  or isinstance(inst, np.ndarray):
+            # data X is epoched 
+            # part of the following code is copied from mne csp
+                n_epochs, n_channels, n_samples = inst.shape
+                self.max_components=n_channels
+              
+                #reshape for filtering
+                X_aux=np.reshape(inst, [n_epochs, n_channels*n_samples])
+                X_s=filter_data(X_aux, self.sampling_freq, **self.filt_params_signal)  
+                
+                #rephase for filtering
+                X_aux=np.reshape(inst, [n_epochs, n_channels*n_samples])
+                X_n=filter_data(X_aux, self.sampling_freq, **self.filt_params_noise)
+                # subtract signal:
+                X_n -= X_s
+                
+                # Estimate single trial covariance
+                #reshape to original shape
+                X_s=np.reshape(X_s, [n_epochs,n_channels,n_samples])
+                self.Xs=X_s
+                covs = np.empty((n_epochs, n_channels, n_channels))
+                for ii, epoch in enumerate(X_s):
```

Comment:
I am wondering if this is really necessary. Why not just use mne.compute_covariance based on the epochs, or at least the approach of concatenating the signals before covariance computation ...

Reply:
Well, I implemented following the idea on the [CSP](https://github.com/mne-tools/mne-python/blob/maint/0.20/mne/decoding/csp.py#L22-L559) implementation (in particular in SPoC implementation). 

I haven't checked on mne epoched data, but on numpy n-dimensional array. I have now added some check-lines to raise a NotImplementedError when input is neither a BaseRaw nor a np.ndimentional array of ndim==3.